[Feature] Add columns and information to the CSV Users Download

bella77-69 opened the pull request:
🤖 Resolves  #13206

👋 Introduction

This PR adds new columns into the Users CSV download. These columns are added to the All-users and Community talent tables. The new columns include information collected from the GC employee profile, and they will appear at the far-right end of the CSV file.

🕵️ Details

The following new columns have been added to the CSV download for both tables:
1. Interest in lateral movement: Empty/Yes/No
2. Target time frame for lateral movement: Empty/Answer as on platform
3. Types of organizations for lateral movement: Empty/Answer as on platform if positive
4. Interest in advancement: Empty/Yes/No
5. Target time frame for advancement: Empty/Answer as on platform
6. Types of organizations for advancement: Empty/Answer as on platform
7. Interest in learning opportunities: Empty/Interchange/Academic program/Peer networking/Professional accreditation
8. Retirement eligibility: Empty/Year as from form
9. Mentorship status: Empty/Answer as on platform
10. Interest in mentorship: Empty/Answer as on platform if positive
11. Interest in executive opportunities: Empty/Yes/No
12. Executive coaching status: Empty/Answer as on platform
13. Interest in executive coaching: Empty/Answer as on platform if positive
14. Next role - Target classification group: Empty/Answer as on platform
15. Next role - Target classification level: Empty/Answer as on platform
16. Next role - Target role: Empty/Answer as on platform
17. Next role - C-suite: Empty/Yes/No
18. Next role - C-suite title: Empty/Answer as on platform
19. Next role - Job title: Empty/Answer as on platform
20. Next role - Functional community: Empty/Answer as on platform
21. Next role - Work streams: Empty/Answer as on platform
22. Next role - Departments: Empty/Answer as on platform
23. Next role - Additional info: Empty/Answer as on platform
24. Career objective - Target classification group: Empty/Answer as on platform
25. Career objective - Target classification level: Empty/Answer as on platform
26. Career objective - Target role: Empty/Answer as on platform
27. Career objective - C-suite: Empty/Yes/No
28. Career objective - C-suite title: Empty/Answer as on platform
29. Career objective - Job title: Empty/Answer as on platform
30. Career objective - Functional community: Empty/Answer as on platform
31. Career objective - Work streams: Empty/Answer as on platform
32. Career objective - Departments: Empty/Answer as on platform
33. Career objective - Additional info: Empty/Answer as on platform
34. About: Empty/Answer as on platform
35. Learning goals: Empty/Answer as on platform
36. How you work best: Empty/Answer as on platform

French:
1. Intérêt pour la mutation latérale: empty/Oui/Non
2. Délai cible pour une mutation latérale: empty/answer as on platform
3. Types d'organisations pour une mutation latérale: empty/ answer as on platform if positive
4. Intérêt pour l'avancement: empty/Oui/Non
5. Délai cible pour l'avancement: empty/ answer as on platform
6. Types d'organisations pour un avancement: empty/ answer as on platform
7. Intérêt pour les possibilités d'apprentissage : empty/Échange/Programme d’études/Réseautage entre pairs/Accréditation professionnelle
8. Admissibilité à la retraite: empty/year as from form
9. État du mentorat: empty/ answer as on platform
10. Intérêt pour les possibilités de mentorat: empty/ answer as on platform if positive
11. Intérêt pour des postes de la direction: empty/Oui/Non
12. État du coaching des cadres: empty/ answer as on platform
13. Intérêt pour le coaching des cadres: empty/ answer as on platform if positive
14. Prochain poste - Groupe de classification visé: empty/ answer as on platform
15. Prochain poste - Niveau de classification visé: empty/ answer as on platform
16. Prochain poste - Type de poste visé: empty/ answer as on platform
17. Prochain poste – Cadre supérieur: empty/Oui/Non
18. Prochain poste - Titre du rôle de cadre supérieur: empty/ answer as on platform
19. Prochain poste - Titre du poste: empty/ answer as on platform
20. Prochain poste - Collectivité fonctionnelle: empty/ answer as on platform
21. Prochain poste – Volets de travail: empty/ answer as on platform
22. Prochain poste - Ministères: empty/ answer as on platform
23. Prochain poste - Renseignements supplémentaires: empty/ answer as on platform
24. Objectif de carrière - Groupe de classification visé: empty/ answer as on platform
25. Objectif de carrière - Niveau de classification visé: empty/ answer as on platform
26. Objectif de carrière - Type de poste visé: empty/ answer as on platform
27. Objectif de carrière - Cadre supérieur: empty/Oui/Non
28. Objectif de carrière - Titre du rôle de cadre supérieur: empty/ answer as on platform
29. Objectif de carrière - Titre du poste: empty/ answer as on platform
30. Objectif de carrière - Collectivité fonctionnelle: empty/ answer as on platform
31. Objectif de carrière - Volets de travail: empty/ answer as on platform
32. Objectif de carrière - Ministères: empty/ answer as on platform
33. Objectif de carrière - Renseignements supplémentaires: empty/ answer as on platform
34. À propos: empty/ answer as on platform
35. Objectifs d’apprentissage: empty/ answer as on platform
36. Comment vous travaillez le mieux: empty/ answer as on platform

🧪 Testing

1. Login as Admin
2. Navigate to `en/admin/users`
3. Select a candidate and download their csv
4. Ensure that all the new columns listed in the details section are present in the CSV
5. repeat steps for `fr/admin/users`

📸 Screenshot

Example users CSV EN: 
[csv_en.csv](https://github.com/user-attachments/files/20092854/csv_en.csv)

Example users CSV FR: 
[csv_fr.csv](https://github.com/user-attachments/files/20092855/csv_fr.csv)

NienkeBr commented:
Hi, few comments:
- In the English csv, could you please change retirement eligibility and mentorship status to sentence case? "Retirement eligibility". "Mentorship status"
- In the French CSV, it doesn't seem to correctly show the French characters. Is this something that could be addressed?
- In both the English and French files, there is a "No" for the "Next role - C-suite" column and the "Career objective - C-suite" column, however, all the other columns are empty, indicating that this users didn't fill out this information. So these columns should then also be empty.

![image](https://github.com/user-attachments/assets/5d06d21c-115a-4fbd-b763-98fee07a7ece)

bella77-69 commented:
[NienkeBr](https://github.com/NienkeBr) I updated the English csv, changed retirement eligibility and mentorship status to sentence case. And I also updated the csv files above ⬆️ (hopefully they show the French characters)
I also refactored so the columns that were showing empty should have data in them now. 😁

esizer commented on api/app/Generators/UserCsvGenerator.php:
```diff
@@ -119,6 +182,43 @@ public function generate(): self
+                    $this->localizeEnum($employeeProfile->career_planning_promotion_move_organization_type, OrganizationTypeInterest::class), // Career planning - Promotion move organization type
+                    $this->localizeEnum($employeeProfile->career_planning_learning_opportunities_interest, LearningOpportunitiesInterest::class), // Career planning - Learning opportunities interest
+                    $employeeProfile->eligible_retirement_year ? $employeeProfile->eligible_retirement_year->format('Y') : '', // Eligible retirement year
+                    $this->localizeEnum($employeeProfile->career_planning_mentorship_status, Mentorship::class), // Career planning - Mentorship statu
```
```suggestion
                    $this->localizeEnumArry($employeeProfile->career_planning_mentorship_status, Mentorship::class), // Career planning - Mentorship status
```

esizer commented on api/app/Generators/UserCsvGenerator.php:
```diff
@@ -119,6 +182,43 @@ public function generate(): self
+                    $this->yesOrNo($employeeProfile->next_role_is_c_suite_role), // Next role - C-suite role
+                    $this->localizeEnum($employeeProfile->next_role_c_suite_role_title, CSuiteRoleTitle::class),
+                    $employeeProfile->next_role_job_title, // Next role - Job title
+                    $employeeProfile->nextRoleCommunity->name[$this->lang], // Next role - Functional community
```
```suggestion
                    $employeeProfile->nextRoleCommunity?->name[$this->lang] ?? '', // Next role - Functional community
```

bella77-69 replied:
👍

esizer commented on api/app/Generators/UserCsvGenerator.php:
```diff
@@ -119,6 +182,43 @@ public function generate(): self
+                    $this->yesOrNo($employeeProfile->career_objective_is_c_suite_role), // Career objective - C-suite role
+                    $this->localizeEnum($employeeProfile->career_objective_c_suite_role_title, CSuiteRoleTitle::class), // Career objective - C-suite role title
+                    $employeeProfile->career_objective_job_title, // Career objective - Job title
+                    $employeeProfile->careerObjectiveCommunity->name[$this->lang], // Career objective - Functional community
```
```suggestion
                    $employeeProfile->careerObjectiveCommunity?->name[$this->lang] ?? '', // Career objective - Functional community
```

bella77-69 replied:
👍

bella77-69 commented on api/app/Generators/UserCsvGenerator.php:
```diff
@@ -90,6 +139,20 @@ public function generate(): self
                     return $userSkill->skill->name[$this->lang] ?? '';
                 });
 
+                $employeeProfile = $user->EmployeeProfile;
```
🙃

NienkeBr commented:
Thanks! My comments have been addressed.

esizer commented on api/app/Generators/UserCsvGenerator.php:
```diff
@@ -119,6 +182,43 @@ public function generate(): self
+                    $this->localizeEnum($employeeProfile->career_planning_promotion_move_organization_type, OrganizationTypeInterest::class), // Career planning - Promotion move organization type
+                    $this->localizeEnum($employeeProfile->career_planning_learning_opportunities_interest, LearningOpportunitiesInterest::class), // Career planning - Learning opportunities interest
+                    $employeeProfile->eligible_retirement_year ? $employeeProfile->eligible_retirement_year->format('Y') : '', // Eligible retirement year
+                    $this->localizeEnum($employeeProfile->career_planning_mentorship_status, Mentorship::class), // Career planning - Mentorship status
```
```suggestion
                    $this->localizeEnumArray($employeeProfile->career_planning_mentorship_status, Mentorship::class), // Career planning - Mentorship status
```

esizer commented on api/app/Generators/UserCsvGenerator.php:
```diff
@@ -119,6 +182,43 @@ public function generate(): self
+                    // new columns
+                    $this->yesOrNo($employeeProfile->career_planning_lateral_move_interest), // Career planning - Lateral move interest
+                    $this->localizeEnum($employeeProfile->career_planning_lateral_move_time_frame, TimeFrame::class), // Career planning - Lateral move time frame
+                    $this->localizeEnum($employeeProfile->career_planning_lateral_move_organization_type, OrganizationTypeInterest::class), // Career planning - Lateral move organization type
```
```suggestion
                    $this->localizeEnumArray($employeeProfile->career_planning_lateral_move_organization_type, OrganizationTypeInterest::class), // Career planning - Lateral move organization type
```

esizer commented on api/app/Generators/UserCsvGenerator.php:
```diff
@@ -119,6 +182,43 @@ public function generate(): self
+                    $this->localizeEnum($employeeProfile->career_planning_lateral_move_organization_type, OrganizationTypeInterest::class), // Career planning - Lateral move organization type
+                    $this->yesOrNo($employeeProfile->career_planning_promotion_move_interest), // Career planning - Promotion move interest
+                    $this->localizeEnum($employeeProfile->career_planning_promotion_move_time_frame, TimeFrame::class), // Career planning - Promotion move time frame
+                    $this->localizeEnum($employeeProfile->career_planning_promotion_move_organization_type, OrganizationTypeInterest::class), // Career planning - Promotion move organization type
```
```suggestion
                    $this->localizeEnumArray($employeeProfile->career_planning_promotion_move_organization_type, OrganizationTypeInterest::class), // Career planning - Promotion move organization type
```

esizer commented on api/app/Generators/UserCsvGenerator.php:
```diff
@@ -119,6 +182,43 @@ public function generate(): self
+                    $this->yesOrNo($employeeProfile->career_planning_promotion_move_interest), // Career planning - Promotion move interest
+                    $this->localizeEnum($employeeProfile->career_planning_promotion_move_time_frame, TimeFrame::class), // Career planning - Promotion move time frame
+                    $this->localizeEnum($employeeProfile->career_planning_promotion_move_organization_type, OrganizationTypeInterest::class), // Career planning - Promotion move organization type
+                    $this->localizeEnum($employeeProfile->career_planning_learning_opportunities_interest, LearningOpportunitiesInterest::class), // Career planning - Learning opportunities interest
```
```suggestion
                    $this->localizeEnumArray($employeeProfile->career_planning_learning_opportunities_interest, LearningOpportunitiesInterest::class), // Career planning - Learning opportunities interest
```

esizer reviewed (changes requested):
Sorry, I noticed a few of the `localizeEnum` need to be updated to `localizeEnumArray` :sweat_smile:
